src/deploy/functions/params.ts

```diff
@@ -0,0 +1,342 @@
+import { logger } from "../../logger";
+import { FirebaseError } from "../../error";
+import { promptOnce } from "../../prompt";
+import * as build from "./build";
+import { assertExhaustive } from "../../functional";
+
+type CEL = build.Expression<string> | build.Expression<number> | build.Expression<boolean>;
+function isCEL(expr: string | number | boolean): expr is CEL {
+  return typeof expr === "string" && expr.includes("{{") && expr.includes("}}");
+}
+function dependenciesCEL(expr: CEL): string[] {
+  return /params\.(\w+)/.exec(expr)?.slice(1) || [];
+}
+
+/**
+ * Resolves a numeric field in a Build to an an actual numeric value.
+ * Fields can be literal or a {{ }} delimited CEL expression referencing param values.
+ * Currently only the CEL identity {{ params.PARAMNAME }} is implemented.
+ */
+export function resolveInt(
+  from: number | build.Expression<number>,
+  paramValues: Record<string, build.Field<string | number | boolean>>
+): number {
+  if (typeof from === "string" && /{{ params\.(\w+) }}/.test(from)) {
+    const match = /{{ params\.(\w+) }}/.exec(from);
+    const referencedParamValue = paramValues[match![1]];
+    if (typeof referencedParamValue !== "number") {
+      throw new FirebaseError(
+        "Referenced numeric parameter '" +
+          match +
+          "' resolved to non-number value " +
+          referencedParamValue
+      );
+    }
+    return referencedParamValue;
+  } else if (typeof from === "string") {
+    throw new FirebaseError("CEL evaluation of expression '" + from + "' not yet supported");
+  }
+  return from;
+}
+
+/**
+ * Resolves a string field in a Build to an an actual string.
+ * Fields can be literal or a {{ }} delimited CEL expression referencing param values.
+ * Currently only the CEL identity {{ params.PARAMNAME }} is implemented.
+ */
+export function resolveString(
+  from: string | build.Expression<string>,
+  paramValues: Record<string, build.Field<string | number | boolean>>
+): string {
```

Comment:
please write the above and below function again to have fewer indentations and be more readable. This is a great example for another reason to value early exit for errors & putting most of the body at main indentation level. Notice that the first if/else if are on LOC 51 & 53. Line 77 is where this condition stops. By the time I see the backup case of "return from" I no longer remember what that's the edge case for.

Reply:
Took a stab at it.